server/src/main/java/org/opensearch/cluster/routing/allocation/AwarenessReplicaBalance.java

```diff
@@ -0,0 +1,112 @@
+
+    private volatile Boolean awarenessBalance;
+
+    public AwarenessReplicaBalance(Settings settings, ClusterSettings clusterSettings) {
+        this.awarenessAttributes = CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING, this::setAwarenessAttributes);
+        setForcedAwarenessAttributes(CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(
+            CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING,
+            this::setForcedAwarenessAttributes
+        );
+        setAwarenessBalance(CLUSTER_ROUTING_ALLOCATION_AWARENESS_BALANCE_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_AWARENESS_BALANCE_SETTING, this::setAwarenessBalance);
+
+    }
+
+    private void setAwarenessBalance(Boolean awarenessBalance) {
+        this.awarenessBalance = awarenessBalance;
+    }
+
+    private void setForcedAwarenessAttributes(Settings forceSettings) {
+        Map<String, List<String>> forcedAwarenessAttributes = new HashMap<>();
+        Map<String, Settings> forceGroups = forceSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : forceGroups.entrySet()) {
+            List<String> aValues = entry.getValue().getAsList("values");
+            if (aValues.size() > 0) {
+                forcedAwarenessAttributes.put(entry.getKey(), aValues);
+            }
+        }
+        this.forcedAwarenessAttributes = forcedAwarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    /*
+    For a cluster having zone as awareness attribute , it will return the size of zones if set it forced awareness attributes
+
+    If there are multiple forced awareness attributes, it will return size of the largest list, as all copies of data
+    is supposed to get distributed amongst those.
+
+    cluster.routing.allocation.awareness.attributes: rack_id , zone
+    cluster.routing.allocation.awareness.force.zone.values: zone1, zone2
+    cluster.routing.allocation.awareness.force.rack_id.values: rack_id1, rack_id2, rack_id3
+
+    In this case,  awareness attributes would be 3.
+     */
+    public int awarenessAttributes() {
+        int awarenessAttributes = 1;
+        if (this.awarenessBalance == false) {
+            return awarenessAttributes;
+        }
+        for (String awarenessAttribute : this.awarenessAttributes) {
+            if (forcedAwarenessAttributes.containsKey(awarenessAttribute)) {
+                awarenessAttributes = max(awarenessAttributes, forcedAwarenessAttributes.get(awarenessAttribute).size());
+            }
+        }
+        return awarenessAttributes;
+    }
+
+    public Predicate<Integer> replicaPredicate = (replicaCount) -> {
```

Comment:
I would recommend having a public method called `validate` preferably throwing a `ValidationException` containing the error properties, expected vs actual

Reply:
done